src/semidiscretization/semidiscretization.jl

```diff
@@ -15,6 +15,26 @@ Return the number of degrees of freedom associated with each scalar variable.
     ndofs(mesh, solver, cache)
 end
 
+"""
+    ndofsglobal(semi::AbstractSemidiscretization)
+
+Return the global number of degrees of freedom associated with each scalar variable.
+"""
+@inline function ndofsglobal(semi::AbstractSemidiscretization)
+    mesh, _, solver, cache = mesh_equations_solver_cache(semi)
+    ndofsglobal(mesh, solver, cache)
+end
+
+"""
+    ndofsglobal(mesh, solver, cache)
+
+Return the global number of degrees of freedom associated with each scalar variable.
+Defaults to ndofs when there is no special implementation for parallel computations.
+"""
+@inline function ndofsglobal(mesh, solver, cache)
```

Comment:
```suggestion
"""
    ndofsglobal(mesh, solver, cache)

Return the global number of degrees of freedom associated with each scalar variable.
Defaults to [`ndofs`](@ref) when there is no special implementation for 
MPI-parallel computations.
"""
@inline function ndofsglobal(mesh, solver, cache)
```

Reply:
I am not sure about it.

My concern was that with `ndofsglobal(semi::AbstractSemidiscretization)` introduced above `ndofsglobal(mesh, solver, cache)` might get called with some mesh, solver, cache combinations for which is does not exist.

If I see it correctly there is only
`@inline function ndofsglobal(mesh, dg::DG, cache)` in `dg.jl`.

Reply:
What's you opinion, @sloede?

Reply:
I'd consider it an implementation detail as well.

However, does it really need a fallback? Or, alternatively, could we check for `mpi_isparallel() == 0` and otherwise error, such that a user cannot accidentally use that with a non-parallelized mesh?

Reply:
> I'd consider it an implementation detail as well.

Ok! However I am afraid I do not understand what this implies.

> However, does it really need a fallback?

I added `ndofsglobal(semi)` to semidiscretization_hyperbolic.jl, which calls `ndofsglobal(mesh, solver, cache)`. So when `solver` is not `DG` there should be a MethodError, shouldn't there?

Reply:
> > I'd consider it an implementation detail as well.
> 
> Ok! However I am afraid I do not understand what this implies.

Our current policy was that if it has a docstring, we consider it part of our API and then changing its API or its behavior would be a breaking change. Thus we don't add docstring to purely internal functions.



> > However, does it really need a fallback?
> 
> I added `ndofsglobal(semi)` to semidiscretization_hyperbolic.jl, which calls `ndofsglobal(mesh, solver, cache)`. So when `solver` is not `DG` there should be a MethodError, shouldn't there?

Right. The question is then whether this is performance critical or if we could add a check that safeguards against misuses. I don't like `ndofsglobal` returning the local number of DOFs if a user just didn't implement anything for their parallel mesh.

As an alternative to the default implementation, you could move the check to the semi implementation such that ndofsglobal is only ever called in parallel simulations.

Reply:
Ah, the docstring! Thanks!

The problem is that I added `ndofsglobal` to 
`function Base.show`
with the idea that always the global number of dofs gets printed, with ndofs as fallback in case there in no `ndofsglobal`.

An alternative could be to introduce a separate function just for printing, say `print_ndofs`, which would then check for `mpi_isparallel()` or have a default fallback.

Reply:
Is this resolved after [b883b3a](https://github.com/trixi-framework/Trixi.jl/pull/1865/commits/b883b3a0490858a28c36a2d4d3415f08bf690316)?

Reply:
It resolves the API thing:

> Our current policy was that if it has a docstring, we consider it part of our API and then changing its API or its behavior would be a breaking change. Thus we don't add docstring to purely internal functions.


Otherwise this PR got stuck on the question of how to print the global number of dofs on Trixi's startup screen for a mesh which does not have a parallel implementation and thus nothing like `ndofsglobal` .

Reply:
I'm a little confused and can't wrap my head around all the implications. Instead of writing back and forth, can we maybe discuss this at the next meeting (if that's not too late for you @benegee)?

Reply:
Would it be a safe way out of this if we were to implement a fallback that checks `mpi_isparallel()` and errors if it is true? That is, if someone implements a mesh that is running in parallel but does not implement this function gets it thrown in the face?